### PR TITLE
Added --ignore-plugin-installation-failure flag to 'cfy snapshot rest…

### PR DESCRIPTION
**Pull request by barucoh**
…ore' command, will ignore any plugins failing to install during a snapshot restore operation.

### CODE REVIEW

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 1]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                    except Exception as ex:
+                        if self._ignore_plugin_installation_failure:
+                            ctx.logger.warning(
+                                'Failed to restore plugin: {0}, '
+                                'ignore-plugin-installation-failure flag '
+                                'used. Proceeding...'.format(plugin))
+                            ctx.logger.debug('Restore plugin failure error: '
+                                             '{0}'.format(ex))
+                            failed_plugins.append(plugin)
+                            plugin_installation_log_message = \
```
Comment (idob-cloudify) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: updating plugin_installation_log_message should not be inside the for loop

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 2]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                        client.plugins.delete(failed_plugin['id'], force=True)
+                    except Exception as ex:
+                        ctx.logger.warning('Failed to delete plugin: {0} as '
+                                           'part of failed plugin upload with '
+                                           'error: {1}. Proceeding...'
+                                           .format(failed_plugin, ex.message))
             finally:
                 os.rmdir(plugins_tmp)
-        ctx.logger.info('Successfully restored plugins')
+        if len(failed_plugins) == 0:
```
Comment (idob-cloudify) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: can use
if failed_plugins
no need "len"

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 3]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                    try:
+                        self._restore_plugin(client,
+                                             tenant,
+                                             plugin,
+                                             plugins_tmp)
+                    except Exception as ex:
+                        if self._ignore_plugin_installation_failure:
+                            ctx.logger.warning(
+                                'Failed to restore plugin: {0}, '
+                                'ignore-plugin-installation-failure flag '
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: I think we don't need to mention that the flag was used

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 1 above)
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: move from the loop

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 4]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                            ctx.logger.debug('Restore plugin failure error: '
+                                             '{0}'.format(ex))
+                            failed_plugins.append(plugin)
+                            plugin_installation_log_message = \
+                                'Plugins: {0} have not been installed ' \
+                                'successfully'.format(failed_plugins)
+                        else:
+                            raise ex
                 self._wait_for_plugin_executions(client)
+                for failed_plugin in failed_plugins:
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: move to function

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 5]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                    # Removing failed plugins from the database and file server
+                    try:
+                        ctx.logger.info('Removing failed plugin footprints.')
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: remove the dot

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 6]
```diff
@@ -587,16 +628,49 @@ def _restore_plugins(self, existing_plugins):
...
+                        client.plugins.delete(failed_plugin['id'], force=True)
+                    except Exception as ex:
+                        ctx.logger.warning('Failed to delete plugin: {0} as '
+                                           'part of failed plugin upload with '
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Not a clear sentence

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 2 above)
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: again...

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 4 above)
Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 5 above)
Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 6 above)
Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 2 above)
Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 7]
```diff
@@ -141,9 +143,47 @@ def restore(self):
             ctx.logger.debug('Removing temp dir: {0}'.format(self._tempdir))
             shutil.rmtree(self._tempdir)
 
+    @staticmethod
+    def __remove_failed_deployments_footprints(tenant_client,
+                                               failed_deployments):
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Add description

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py [hunk 8]
```diff
@@ -218,9 +268,9 @@ def _restore_certificate(self):
 
         if not os.path.exists(
                 os.path.join(archive_cert_dir, INTERNAL_CA_CERT_FILENAME)):
-            for source, target in [
-                    (INTERNAL_CERT_FILENAME, INTERNAL_CA_CERT_FILENAME),
-                    (INTERNAL_KEY_FILENAME, INTERNAL_CA_KEY_FILENAME)]:
+            for source, target in \
```
Comment (adigrabow) [inline workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: It was better before: `in [ `

File: tests/integration_tests/tests/agentless_tests/test_snapshot.py [hunk 9]
```diff
@@ -425,6 +485,53 @@ def _upload_and_restore_snapshot(self,
         # function _add_restart_command.
         time.sleep(5)
 
+    def _upload_and_restore_snapshot_skip_bad_plugin_failure(
```
Comment (adigrabow) [inline tests/integration_tests/tests/agentless_tests/test_snapshot.py]: Use `_upload_and_restore_snapshot`

File: tests/integration_tests/tests/agentless_tests/test_snapshot.py [hunk 10]
```diff
@@ -425,6 +485,53 @@ def _upload_and_restore_snapshot(self,
...
+            self.logger.error('Execution error: {0}'.format(execution.error))
+        self.assertEqual(Execution.TERMINATED, execution.status)
+
+        # Wait for the restart of cloudify-restservice to end.
+        # If the rest-security.conf file was changed the service is being
+        # restarted. As you can see in snapshot_restore.py when calling the
+        # function _add_restart_command.
+        time.sleep(5)
+
+    def _upload_and_restore_snapshot_failure(self,
```
Comment (adigrabow) [inline tests/integration_tests/tests/agentless_tests/test_snapshot.py]: use `_upload_and_restore_snapshot` instead

Comment (barucoh) [reply tests/integration_tests/tests/agentless_tests/test_snapshot.py]: Fixed

File: tests/integration_tests/tests/agentless_tests/test_snapshot.py (hunk 9 above)
Comment (barucoh) [reply tests/integration_tests/tests/agentless_tests/test_snapshot.py]: Fixed

File: workflows/cloudify_system_workflows/snapshots/snapshot_restore.py (hunk 7 above)
Comment (barucoh) [reply workflows/cloudify_system_workflows/snapshots/snapshot_restore.py]: Fixed

**Pull request merged**